### ISSUE TITLE
Token authentication with expiring token.

### ISSUE THREAD

**Issue by itzmanish** [apragacz/django-rest-registration#84]
It would be a nice feature if rest-registration provides inbuilt token authentication with expiring token. 
JWT provides expiring tokens but handling logout with blacklisting token is a headache that I feel but in the case of expiring token, it doesn't need any blacklisting on logout.
It will increase the strength of authentication with the token mechanism.

**Comment by apragacz** [comment]
This is certainly a nice idea to have a "crypto token", whether it would be based on JWT or another signing mechanism. Unfortunately at that moment I have to give it a low priority as I don't have much time so I need to focus on other stuff.

**Comment by itzmanish** [comment]
I can take this up. You only need to review once.

**Comment by apragacz** [comment]
That would be helpful, but before you start doing any work I would like to discuss this first as this feature seems to be non-trivial. This way you don't waste your time on coding a solution which would be either completely refactored or rejected. Will sent some details in next few days.

**Comment by itzmanish** [comment]
OK. If you use telegram then it would be better we talk over telegram.
Username-itzmanish

**Comment by apragacz** [comment]
For the sake of transparency I'd prefer to discuss it here (as it is described in recently added [Contributing Guidelines](https://github.com/apragacz/django-rest-registration/blob/master/CONTRIBUTING.md#pull-requests)). I shall give your an update as soon as possible.

**Comment by apragacz** [comment]
Before we introduce expiring tokens, we need to provide a way to abstract the token generation / revoking process.

First think I want to assume is that specific instance should be able to use at most one method. In other words the user of the library should be able either token set in DB or expiring token, but not both at the same time.

This is certainly a limitation, but with that we can simplify the configuration and the implementation. Also, with accord to the YAGNI principle, I don't think this is required by any user ATM.

Currently I see following options to solve this abstraction:

## Proposal 1

Add these setting keys to `REST_REGISTRATION` dict:
-   `'TOKEN_GENERATOR'`
-   `'TOKEN_REVOKER'`

Which represent the token generating function and token revoking function, accordingly (in case of "crypto token" we wouldn't). They would be used only when `should_retrieve_token` (in `rest_registration.api.views.login`) returns `True`.

advantages:
-   very simple solution
-   we can have simple functions defining the behavior of token generation

disadvantages
-   if the user misuses these settings, or forgets to add matching auth class to `REST_FRAMEWORK['DEFAULT_AUTHENTICATION_CLASSES']` he may end up into trouble and it may be hard to debug the problem (we cannot provide any checks)

## Proposal 2

Add `'TOKEN_MANAGER_CLASS'` setting key to `REST_REGISTRATION` dict which would represent a subclass of following class:

```python
from django.contrib.auth.models import AbstractBaseUser


class AbstractTokenManager:

    def get_auth_class(self):
        raise NotImplementError()

    def provide_token(self, user: AbstractBaseUser) -> str:
        raise NotImplementError()

    def revoke_token(self, user: AbstractBaseUser, token : str = None):
        # may not be always implemented
        raise NotImplementError()
```

By default, we would have `REST_REGISTRATION['TOKEN_MANAGER_CLASS'] = 'rest_registration.token_managers.RestFrameworkTokenManager'` with `RestFrameworkTokenManager` defined as follows:

```python
from django.contrib.auth.models import AbstractBaseUser
from rest_framework.authentication import TokenAuthentication
from rest_framework.authtoken.models import Token


class RestFrameworkTokenManager(AbstractTokenManager):

    def get_auth_class(self):
        return TokenAuthentication

    def provide_token(self, user: AbstractBaseUser) -> str:
        token, _ = Token.objects.get_or_create(user=user)
        return token

    def revoke_token(self, user: AbstractBaseUser, token : str = None):
        try:
            user.auth_token.delete()
        except Token.DoesNotExist:
            raise ValueError("Cannot remove non-existent token")
```

(this is sample implementation, may need refinement!)

With this solution, we could match specific token manager to the DRF auth class, so if user forgets to add it to settings we may detect that with Django system check.

advantages:
-   one setting key
-   we can easily test that the configuration is set properly

disadvantages:
-   complex solution
-   requires using class

I'm aiming towards proposal 2 at this point.

I also think we should create separate issue for this, and it should be blocking this one.

@itzmanish what do you think? Which of these proposals do you prefer? Would you like to implement this abstraction, or just implement a specific expiring token solution (for instance, JWT-based)?

**Comment by itzmanish** [comment]
@apragacz I think Proposal 2 is better solution. However we also targeted to switch to class based views from function based views. From my point of view i see expiring token is much better than any third party JWT token or any other because by using expiring token we can keep it more clean and it will increase compatibility with DRF. 
The biggest problem with JWT I encounter is Token blacklisting and for every login JWT creates new token so managing login session is very difficult. But with expiring token one can even get the feature of JWT with addition to active session management of a user(Whether a user should be logged in one device or multiple device.)
I hope we soon switch to class based view and get Expiring token authentication out of the box.

Or

I have another solution why don't we keep Expiring token as another library so if any one want expiring token in their application they can use it as plug and use. This will keep DRR less complex and small with core features.(However this is just my view. What you say about this?)

**Comment by apragacz** [comment]
> However we also targeted to switch to class based views from function based views.
> ...
> I hope we soon switch to class based view and get Expiring token authentication out of the box.

I'm not sure how adding custom token authentication (assuming we will add custom token retrieval/revoking) has to do with class-based views. You can set the class globally both for CBV and `api_view` decorated views (and set the auth per CBV/decorated view as well): https://www.django-rest-framework.org/api-guide/authentication/#setting-the-authentication-scheme 

Is there is some existing mechanism for token retrieval in DRF? If yes, could you link to the docs / source code?

> From my point of view i see expiring token is much better than any third party JWT token or any other because by using expiring token we can keep it more clean and it will increase compatibility with DRF.
> The biggest problem with JWT I encounter is Token blacklisting and for every login JWT creates new token so managing login session is very difficult. But with expiring token one can even get the feature of JWT with addition to active session management of a user(Whether a user should be logged in one device or multiple device.)
> ...
> I have another solution why don't we keep Expiring token as another library so if any one want expiring token in their application they can use it as plug and use. This will keep DRR less complex and small with core features.(However this is just my view. What you say about this?)

I'm not exactly getting what your imagined expiring token would work.
However I think keeping expiring token as a separate library is a great idea!

Therefore I will implement the abstraction (proposal II) in #98. After that I will close this issue as it should be implemented as a separate library. For now we should consider this issue as blocked by issue #98.

**Comment by itzmanish** [comment]
> I'm not sure how adding custom token authentication (assuming we will add custom token retrieval/revoking) has to do with class-based views. 

So that anyone can extend it for their specific needs.

> Is there is some existing mechanism for token retrieval in DRF? 

Maybe i am wrong but is not (https://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication) provides the same feature of token retrieving and token generations?

> I'm not exactly getting what your imagined expiring token would work.

I don't know. I think after implementation i can tell you what i wanted to say(And the thing is my English is not that good so that i can explain to you about my point of view. Better let this issue solved first).

**Comment by apragacz** [comment]
@itzmanish I'm planning to merge the proposed solution (described in issue #98) which is the merge request #99 soon (withing few days).

If you have any comments please share them now, as changing the spec later may be difficult (especially if it introduces breaking changes).

**Comment by itzmanish** [comment]
> @itzmanish I'm planning to merge the proposed solution (described in issue #98) which is the merge request #99 soon (withing few days).
> 
> If you have any comments please share them now, as changing the spec later may be difficult (especially if it introduces breaking changes).

Seems merge #99 does the desired work and perfectly fine.

**Comment by apragacz** [comment]
Closing on behalf of #98 and #99.
